cuda/matrix/ell_kernels.cu

```diff
@@ -405,6 +407,190 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
     GKO_DECLARE_ELL_CONVERT_TO_DENSE_KERNEL);
 
 
+namespace kernel {
+
+
+template <typename ValueType, typename IndexType>
+__global__ __launch_bounds__(default_block_size) void count_nnz_per_row(
+    size_type num_rows, size_type max_nnz_per_row, size_type stride,
+    const ValueType *__restrict__ values, IndexType *__restrict__ result)
+{
+    constexpr auto warp_size = cuda_config::warp_size;
+    const auto tidx = threadIdx.x + blockIdx.x * blockDim.x;
+    const auto row_idx = tidx / warp_size;
+
+    if (row_idx < num_rows) {
+        IndexType part_result{};
+        for (auto i = threadIdx.x % warp_size; i < max_nnz_per_row;
+             i += warp_size) {
+            if (values[stride * i + row_idx] != zero<ValueType>()) {
+                part_result += 1;
+            }
+        }
+
+        auto warp_tile =
+            group::tiled_partition<warp_size>(group::this_thread_block());
+        result[row_idx] = reduce(
+            warp_tile, part_result,
+            [](const size_type &a, const size_type &b) { return a + b; });
+    }
+}
+
+
+template <typename ValueType, typename IndexType>
+__global__ __launch_bounds__(default_block_size) void fill_in_csr(
+    size_type num_rows, size_type max_nnz_per_row, size_type stride,
+    const ValueType *__restrict__ source_values,
+    const IndexType *__restrict__ source_col_idxs,
+    IndexType *__restrict__ result_row_ptrs,
+    IndexType *__restrict__ result_col_idxs,
+    ValueType *__restrict__ result_values)
+{
+    const auto tidx = threadIdx.x + blockDim.x * blockIdx.x;
+
+    if (tidx < num_rows) {
+        auto write_to = result_row_ptrs[tidx];
+        for (auto i = 0; i < max_nnz_per_row; i++) {
+            if (source_values[tidx + stride * i] != zero<ValueType>()) {
```

Comment:
Here, you could also save the index calculation inside a separate variable (the result of `tidx + stride * i`).